### PR TITLE
Add FrameBufferCubemap class.

### PR DESCRIPTION
**Pull request by realitix**
This class allows to render in a cubemap framebuffer easily.

It's very useful for point light shadowing.
It works very well and it's documented.

### CODE REVIEW

**Comment by realitix** [comment]
I have to rewrite this by update Framebuffer.

**Comment by realitix** [comment]
Ok I updated it. It extends FrameBuffer now.

**Comment by xoppa** [comment]
Thanks! I don't think that a framebuffer class should be controlling (or be aware of) the camera. Should be more flexible and just a few loc to leave this up to the caller. Having a set/get/nextSide should be enough. Besides that it looks good to me.

Btw, it is very unlikely that the caller actually ever needs a full blown Cubemap with 6 Pixmaps, thats just wasting memory. The same goes for the FrameBuffer class, of which many users probably never actually will need the Pixmap.

**Comment by realitix** [comment]
I understand what you said but this part of the code:

```
case NegativeX:
+ camera.up.set(0, -1, 0);
+ camera.direction.set(-1, 0, 0);
+ break;
+ case NegativeY:
+ camera.up.set(0, 0, -1);
+ camera.direction.set(0, -1, 0);
+ break;
+ case NegativeZ:
+ camera.up.set(0, -1, 0);
+ camera.direction.set(0, 0, -1);
+ break;
+ case PositiveX:
+ camera.up.set(0, -1, 0);
+ camera.direction.set(1, 0, 0);
+ break;
+ case PositiveY:
+ camera.up.set(0, 0, 1);
+ camera.direction.set(0, 1, 0);
+ break;
+ case PositiveZ:
+ camera.up.set(0, -1, 0);
+ camera.direction.set(0, 0, 1);
+ break;
+ default:
+ break;
+ }

```

has to be in the framebuffer or somewhere because it's required to get a good cubemap. Maybe a FramebufferCubemapHelper. I don't know where you would see this part of code but it has to be somewhere.

**Comment by xoppa** [comment]
Although not *required to get a good cubemap* (it's perfectly fine to use whatever projection you need), I agree that mapping a cubemap side to an orientation is useful. However, that isn't the responsibility of (nor should it be restricted to) the framebuffer. The framebuffer is a render target, it should not care about projection.

The orientation is a property of the side (hence the naming). So I guess that would be a good place to put such value. So the caller could do for example:

    cam.up.set(side.up);
    cam.direction.set(side.direction);
    cam.update();

**Comment by realitix** [comment]
Ok, smart.

**Comment by realitix** [comment]
Xoppa, I updated the framebufferCubemap. I hope this one is the good ;-)

File: gdx/src/com/badlogic/gdx/graphics/Cubemap.java [hunk 1]
```diff
@@ -43,32 +44,49 @@
...
+			this.direction = new Vector3(directionX, directionY, directionZ);
 		}
 
 		/** @return The OpenGL target (used for glTexImage2D) of the side. */
 		public int getGLEnum () {
 			return glEnum;
 		}
+
+		/** @return The up vector of the side. */
+		public Vector3 getUp () {
```
Comment (xoppa) [inline gdx/src/com/badlogic/gdx/graphics/Cubemap.java]: Could you change this and the direction getter to:

    public Vector3 getUp (Vector3 out) { return out.set(up); }

**Comment by xoppa** [comment]
Can you explain (preferable with a link to the spec) why you're using negative Y for the up vector? I have a feeling that you be compensating for the flipping v issue.

I'm not quite sure about casting the GLTexture to the Texture. This kinda exposes an API issue, e.g. one could call `FrameBufferCubemap#getColorBufferTexture()` (which is perfectly valid api-wise) and get a cast exception. Perhaps we should use generics instead. On the other hand that would complicate the API and might break existing code.

I guess it makes sense to add a generic GLFramebuffer class (just like there's a GLTexture) which represents an OpenGL framebuffer without specific implementation. And then let FrameBuffer (`GLFrameBuffer<Texture>`) and CubemapFrameBuffer (`GLFrameBuffer<Cubemap>`) be specific implementations of that class.

**Comment by realitix** [comment]
OK it's a good idea, I will commit a first version with this concept.
Just to be sure, I create the GLFrameBuffer<T> class and then I extend it:
FrameBuffer extends GLFrameBuffer<Texture>
FrameBufferCubemap extends GLFrameBuffer<Cubemap>

Is this what you mean ?

**Comment by realitix** [comment]
FrameBuffer extends GLFrameBuffer \<Texture\>
FrameBufferCubemap extends GLFrameBuffer \<Cubemap\>

Markdown escape....

**Comment by xoppa** [comment]
Sorry for the delay. Yes, that's indeed what I meant.

**Comment by realitix** [comment]
Ok Xoppa, it's done.
I wait for your comment.
I tested it it works well.
I us Negative Y because Positive Y does not work but I can't explain the reason.

**Comment by xoppa** [comment]
Looks good. The negative Y needs some substantiation though. Can you add your test? I will then pull in locally to test and merge.

**Comment by realitix** [comment]
Ok Xoppa, I added tests.

**Comment by xoppa** [comment]
Looks like the test doesn't run (the Cubemap is never constructed), can you update it?

**Comment by xoppa** [comment]
The test throws at GL_FRAMEBUFFER_INCOMPLETE_MISSING_ATTACHMENT, any clue what might be wrong? https://github.com/xoppa/libgdx/tree/pr2933

**Comment by realitix** [comment]
I experienced this before. Try adding the depth buffer by passing true in the framebuffercubemap constructor.

**Comment by realitix** [comment]
I updated it because there were new conflict with Cubemap

**Comment by realitix** [comment]
I updated tests to create depth buffer, it may be the source of your problem. Can you try again ?

**Comment by xoppa** [comment]
Yes using a depth buffer it works.

**Comment by realitix** [comment]
It's a hardware bug. On my computer it works without depth buffer

**Comment by xoppa** [comment]
Sorry for the delay. Need some more time to check some things, but overall I think its good. No need to update the PR after changes, I keep it updated in a separate branch.

**Comment by realitix** [comment]
Ok Xoppa.
New conflicts but I let you manage it.

**Comment by realitix** [comment]
Hello Xoppa,
Did you have time to watch at this FramebufferCubemap ?

**Comment by xoppa** [comment]
Sorry for the delay. It's merged now.

**Comment by realitix** [comment]
Great thanks !

**Pull request closed without merging**